Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/README.md

```diff
@@ -0,0 +1,106 @@
+It would be worthwhile working through that repository first to understand the concepts before going through ours.  
+The Pytorch model including the extra files such as the vocabulary file and class mapping file, which are essential to make the model functional,
+are saved to a MLflow tracking server using the function `mlflow.pytorch.log_model`. This requires a tracking server to be set up with a sql backend before starting.
+
+
+By default,  the script exports the model file as `model.pth`.
+This example workflow includes the following steps,
+1. A pre trained Hugging Face bert model is fine-tuned to classify news.
+2. Model is saved with extra files model,summary, parameters and extra files at the end of training to MLflow server
+3. Deployment of the  model in a Torchserve Docker image.
+
+Torchserve deployment plugin has the ability to detect and add the `requirements.txt` and the extra files. And hence, during the
+mar file generation, TorchServe automatically bundles the `requirements.txt`and extra files along with the model.
+
+## MLflow server
+To allow the code to log model artifacts remotely, a tracking server needs to be set up with a 
+sql [backend store](https://www.mlflow.org/docs/latest/tracking.html#backend-stores). 
+Not required but advisable is setting up a bucket based [artifact store](https://www.mlflow.org/docs/latest/tracking.html#artifact-stores)
+
+To configure MLflow to log to a custom (non-default) tracking location, set the MLFLOW_TRACKING_URI environment variable, e.g. via export MLFLOW_TRACKING_URI=http://localhost:5000/.
+For more details, see [the docs](https://mlflow.org/docs/latest/tracking.html#where-runs-are-recorded).
+
+Our MLflow server used [Google Cloud Storage](https://www.mlflow.org/docs/latest/tracking.html#google-cloud-storage) as its artifact store
+and so requires a [json credentials file][https://cloud.google.com/docs/authentication/getting-started] for a 
+service account with access to the bucket. Set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to the location of this file.  
+This is because although we track metrics and metadata about the
+model in the tracking server backend store, the artifacts are sent directly to the cloud bucket.
+
+## Package Requirement
+
+Install MLflow per the [instructions](https://github.com/mlflow/mlflow#installing).  
+This project doesn't require `conda` on the `PATH`, but does require `docker` to be installed.
+
+## Training
+### Building the trainer image
+To run the code reproducibly, we use the `docker` project environment.
+See the [documentation](https://www.mlflow.org/docs/latest/projects.html#project-environments).  
+
+The image contains the required GCP credentials for logging models. This is passed as a build argument using base64 encoding.
+
+To build the image run
+```
+docker build --build-arg GCP_CREDS_JSON_BASE64="$(base64 $GOOGLE_APPLICATION_CREDENTIALS)" -f ./Dockerfile-trainer .
```

Comment:
Think you already know that, but just remember to check again the Dockerfile name after clean up PR!